utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use zerovec::ule::AsULE;
+use zerovec::ZeroVec;
+
+// Match-node lead unit values, after masking off intermediate-value bits:
+
+// 00..0f: Branch node. If node!=0 then the length is node+1, otherwise
+// the length is one more than the next byte.
+
+// For a branch sub-node with at most this many entries, we drop down
+// to a linear search.
+const MAX_BRANCH_LINEAR_SUB_NODE_LENGTH: usize = 5;
+
+// 0030..003f: Linear-match node, match 1..16 units and continue reading the next node.
+const MIN_LINEAR_MATCH: u16 = 0x30;
+const MAX_LINEAR_MATCH_LENGTH: u16 = 0x10;
+
+// Match-node lead unit bits 14..6 for the optional intermediate value.
+// If these bits are 0, then there is no intermediate value.
+// Otherwise, see the *NodeValue* constants below.
+const MIN_VALUE_LEAD: u16 = MIN_LINEAR_MATCH + MAX_LINEAR_MATCH_LENGTH; // 0x40
+const NODE_TYPE_MASK: u16 = MIN_VALUE_LEAD - 1; // 0x003f
+
+// A final-value node has bit 15 set.
+const VALUE_IS_FINAL: u16 = 0x8000;
+
+// Compact value: After testing bit 0, shift right by 15 and then use the following thresholds.
+const MAX_ONE_UNIT_VALUE: u16 = 0x3fff;
+
+const MIN_TWO_UNIT_VALUE_LEAD: u16 = MAX_ONE_UNIT_VALUE + 1; // 0x4000
+
+const MAX_ONE_UNIT_NODE_VALUE: u16 = 0xff;
+
+const MIN_TWO_UNIT_NODE_VALUE_LEAD: u16 = MIN_VALUE_LEAD + ((MAX_ONE_UNIT_NODE_VALUE + 1) << 6); // 0x4040
+
+const THREE_UNIT_NODE_VALUE_LEAD: u16 = 0x7fc0;
+
+const THREE_UNIT_VALUE_LEAD: u16 = 0x7fff;
+
+// Compact delta integers.
+const MAX_ONE_UNIT_DELTA: u16 = 0xfbff;
+const MIN_TWO_UNIT_DELTA_LEAD: u16 = MAX_ONE_UNIT_DELTA + 1; // 0xfc00
+const THREE_UNIT_DELTA_LEAD: u16 = 0xffff;
+
+fn skip_value(pos: usize, lead: u16) -> usize {
+    if lead < MIN_TWO_UNIT_VALUE_LEAD {
+        pos
+    } else if lead < THREE_UNIT_VALUE_LEAD {
+        pos + 1
+    } else {
+        pos + 2
+    }
+}
+
+fn skip_node_value(pos: usize, lead: u16) -> usize {
+    if lead < MIN_TWO_UNIT_NODE_VALUE_LEAD {
+        pos
+    } else if lead < THREE_UNIT_NODE_VALUE_LEAD {
+        pos + 1
+    } else {
+        pos + 2
+    }
+}
+
+/// This struct represents a de-serialized Char16Trie that was exported from
```

Comment:
Way too much information for public API docs. For public docs, all that a user should care about is that the data is an array of 16-bit units, and it encodes a structure for mapping strings/sequences to values. That is, the documentation that is on the class top level here: https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html

What you have here is from the private/internal details that explain the nitty-gritty details to someone trying to understand the code and the data structure. It should be here somewhere, but it should remain private/internal.

Reply:
It's good to have these docs but perhaps you can attach them to an internal module or somewhere like that.